src/Icicle/Avalanche/Prim/Compounds.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE RecordWildCards #-}
+module Icicle.Avalanche.Prim.Compounds (
+    FlatOps(..)
+  , flatOps
+  ) where
+
+import              Icicle.Common.Type
+import              Icicle.Common.Exp
+import qualified    Icicle.Common.Exp.Prim.Minimal as Min
+import              Icicle.Avalanche.Prim.Flat
+
+import              P
+
+type X a n = Exp a n Prim
+
+data FlatOps a n = FlatOps {
+    arrLen  :: ValType -> X a n -> X a n
+  , arrIx   :: ValType -> X a n -> X a n -> X a n
+  , arrNew  :: ValType -> X a n -> X a n
+  , arrUpd  :: ValType -> X a n -> X a n -> X a n -> X a n
+  , arrZip  :: ValType -> ValType
+            -> X a n -> X a n -> X a n
+
+  , mapPack :: ValType -> ValType -> X a n -> X a n -> X a n
+  , mapKeys :: ValType -> ValType -> X a n -> X a n
+  , mapVals :: ValType -> ValType -> X a n -> X a n
+
+  , bufPush :: Int -> ValType -> X a n -> X a n -> X a n
+  , bufRead :: Int -> ValType -> X a n -> X a n
+
+  , isSome   :: ValType -> X a n -> X a n
+  , optionGet:: ValType -> X a n -> X a n
+  , someF    :: ValType -> X a n -> X a n
+
+  , isRightF :: ValType -> ValType -> X a n -> X a n
+  , left     :: ValType -> ValType -> X a n -> X a n
+  , right    :: ValType -> ValType -> X a n -> X a n
+
+  , fstF     :: ValType -> ValType -> X a n -> X a n
+  , sndF     :: ValType -> ValType -> X a n -> X a n
+  , pair     :: ValType -> ValType -> X a n -> X a n -> X a n
+
+  , structGet:: StructField -> ValType -> StructType
+             -> X a n -> X a n
+
+  , relEq    :: ValType -> X a n -> X a n -> X a n
+
+  }
+
+flatOps :: a -> FlatOps a n
+flatOps a_fresh
+ = FlatOps{..}
```

Comment:
record wildcard probably not necessary here, and could be written just as cleanly without it I guess